### ISSUE TITLE
cockpitfslist return extra attributes

### ISSUE THREAD

**Issue by leomoty** [cockpit-project/cockpit#14170]
Would it be possible for cockpitfslist to return a few extra attributes? I want to create a module for cockpit but this is already pretty close as far as listing files goes.

Like file size, last modified, owner, group and so forth? I can try to implement the change myself if that would be an acceptable compromise.

**Comment by martinpitt** [comment]
fslist1 essentially wraps [GFileMonitor](https://developer.gnome.org/gio/stable/GFileMonitor.html), and neither that nor the underlying Linux API  [inotify](https://www.man7.org/linux/man-pages/man7/inotify.7.html) give you the stat data "for free". [GFileEnumerator](https://developer.gnome.org/gio/stable/GFileEnumerator.html) does, I think, but it should be careful to not always do extra stat() calls on each file unless the user requested it.

So, what could work is to add a new `stat` boolean option to the fslist1 control channel that enables stat information. That is certainly more efficient than calling `ls` or `stat` via a spawn channel all the time.

You just need to be aware that your page will need to have a tight cockpit version requirement in its manifest, to make sure the new API is available. Thus it will not work on e. g. existing RHEL/CentOS 7 or even recent RHEL 8 or Debian/Ubuntu stable releases. Of course over time the API will be more widely available.

So TL/DR, if you want to work on this, I think it's a good idea :grinning:

**Comment by leomoty** [comment]
Hey @martinpitt, are you sure? it seems to be using `g_file_enumerate_children_async`, and the JSON message is constructed using `GFileInfo`?

Isn't it already a GFileEnumerator?

https://github.com/cockpit-project/cockpit/blob/master/src/bridge/cockpitfslist.c#L126

**Comment by leomoty** [comment]
And regarding the API support, not a big deal for now, it is just a side project for fun, if it takes a while, well it takes a while, for my own consumption I can always build bridge myself.

**Comment by martinpitt** [comment]
@leomoty : Yes, that's exactly what I meant; I can't say off-hand whether calling more g_file_info_get_*() methods (to get the extra attributes that you want) will do more stat() calls -- my gut feeling is that it will, as just for the name you don't need stat() at all. If that's the case, then I'd rather make that opt-in.

**Comment by leomoty** [comment]
Hey @martinpitt, I am not too familiar with glib and such, so yesterday I started testing it, it seems that  you need to call `g_file_enumerate_children_async` with the extra attributes you wanna retrieve.

For example, if I want to grab the owner, I need to append this `G_FILE_ATTRIBUTE_OWNER_USER`: to:

```c
  g_file_enumerate_children_async (file,
                                   G_FILE_ATTRIBUTE_STANDARD_NAME "," G_FILE_ATTRIBUTE_STANDARD_TYPE, // append it here
                                   G_FILE_QUERY_INFO_NONE,
                                   G_PRIORITY_DEFAULT,
                                   self->cancellable,
                                   on_enumerator_ready,
                                   g_object_ref (self));
```

If that is not done, calling `g_file_info_get_*` will with that property will yield `null` as a result. Nonetheless, might as well keep it as optional.

Not directly related but, I asked on IRC:

What is the easiest way to run my compiled cockpit-bridge? What I have been trying right now is, run `make install` and login again into a new session to see the changes.

**Comment by martinpitt** [comment]
@leomoty : Right, and my suspicion is that these extra flags will cause additional stat() calls. But how about you just add them, plus an unit test, then it's much easier to compare master with your PR, including strace'ing it and seeing how much extra stuff it does. Once it outputs what you need, we can then fine-tune it.

> What is the easiest way to run my compiled cockpit-bridge

If you want to test it right on your dev system (not test/image-prepare), there's two ways:

    sudo mount -o bind `pwd`/cockpit-bridge /usr/bin/bridge

This requires root, but avoids *any* file system changes, and is very robust. If you don't want/have root, you can do something like

    ln -s `pwd`/cockpit-bridge ~/bin

and make sure that in `$PATH` your ~/bin/ comes before /usr/bin/. If you log into cockpit, you can then check with `pgrep -a` and `ls -l /proc/<pid>/exe` that it uses the right bridge.

**Comment by martinpitt** [comment]
This was implemented in PR #14255, so I am closing this now. Thanks!
